src/main/kotlin/no/skatteetaten/aurora/boober/service/resourceprovisioning/S3StorageGridProvisioner.kt

```diff
@@ -0,0 +1,211 @@
+package no.skatteetaten.aurora.boober.service.resourceprovisioning
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+import com.fkorotkov.kubernetes.newObjectMeta
+import io.fabric8.kubernetes.api.model.HasMetadata
+import mu.KotlinLogging
+import no.skatteetaten.aurora.boober.feature.*
+import no.skatteetaten.aurora.boober.model.AuroraDeploymentSpec
+import no.skatteetaten.aurora.boober.model.openshift.*
+import no.skatteetaten.aurora.boober.service.*
+import no.skatteetaten.aurora.boober.service.openshift.OpenShiftClient
+import no.skatteetaten.aurora.boober.service.openshift.OpenShiftDeployer
+import no.skatteetaten.aurora.boober.service.openshift.OpenShiftResponse
+import no.skatteetaten.aurora.boober.service.openshift.OperationType
+import no.skatteetaten.aurora.boober.utils.convert
+import no.skatteetaten.aurora.boober.utils.normalizeLabels
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Service
+
+data class ObjectAreaWithCredentials(
+    val objectArea: S3ObjectArea,
+    val s3Credentials: StorageGridCredentials
+)
+
+data class S3ObjectArea(
+    val tenant: String,
+    val bucketName: String,
+    val area: String,
+    val specifiedAreaKey: String = area
+)
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+data class StorageGridCredentials(
+    val tenantName: String,
+    val serviceEndpoint: String,
+    val accessKey: String,
+    val secretKey: String,
+    val bucketName: String,
+    val objectPrefix: String,
+    val username: String,
+    val password: String,
+    val bucketRegion: String? = null
+)
+
+class S3ProvisioningException(message: String, cause: Throwable? = null) : ProvisioningException(message, cause)
+
+private val logger = KotlinLogging.logger { }
+
+@Service
+class S3StorageGridProvisioner(
+    val openShiftDeployer: OpenShiftDeployer,
+    val openShiftClient: OpenShiftClient,
+    val openShiftCommandService: OpenShiftCommandService,
+    val herkimerService: HerkimerService,
+    val operationScopeFeature: OperationScopeFeature,
+    @Value("\${minio.bucket.region:us-east-1}") val defaultBucketRegion: String,
+    @Value("\${storagegrid.provisioning-timeout:20000}") val provisioningTimeout: Long,
+    @Value("\${storagegrid.provisioning-status-check-interval:1000}") val statusCheckIntervalMillis: Long
+) {
+    fun getOrProvisionCredentials(adc: AuroraDeploymentSpec): List<ObjectAreaWithCredentials> {
+
+        provisionMissingObjectAreas(adc)
+        return herkimerService.getObjectAreaAdminCredentials(adc)
+            .also { logger.debug("Found ${it.size} ObjectArea credentials for ApplicationDeployment ${adc.namespace}/${adc.name}") }
+    }
+
+    fun provisionMissingObjectAreas(adc: AuroraDeploymentSpec) {
+        val existingResources = herkimerService.getObjectAreaResourcesIndex(adc)
```

Comment:
Er det en god grunn for at boober trenger å gjore dette? Jeg synes at boober kun skal bry seg om å skrive CR. Operatoren håndterer med å returnere hvis det allerede finnes i herkimer.

Reply:
hmm... jeg er ikke uenig. Så lenge vi får til en ordentlig merge med den eksisterende ressursen vil det gå greit. Kan fikse det 👍